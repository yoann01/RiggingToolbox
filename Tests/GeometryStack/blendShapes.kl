
require RiggingToolbox;

operator entry(){

  StartFabricProfiling();

  FilePath basePath = FilePath("${FABRIC_RIGGINGTOOLBOX_PATH}/Tests/GeometryStack/Resources/").expandEnvVars();

  String json = "    {\n\
    \"geomOperators\": [
      {\n\
       \"type\": \"AlembicGeometryGenerator\",\n\
       \"filePath\": \"sphereBlendShapes.abc\",\n\
       \"geometryNames\": [\"SphereRefShape\"]
      },
      {\n\
       \"type\": \"BlendShapesModifier\",\n\
       \"filePath\": \"sphereBlendShapes.abc\",\n\
       \"referenceGeometryName\": \"SphereRefShape\",\n\
       \"targetGeometryNames\": [\"SphereRefTarget*\"]\n\
       \"displayDebugging\": true\n\
      }
    ]
  }";

  GeometryStack stack();
  PersistenceContext persistenceContext();
  persistenceContext.filePath = basePath.string();
  stack.loadJSONString(persistenceContext, json);

  BlendShapesModifier blendShapesModifier = stack.getGeometryOperator(1);

  EvalContext context();
  stack.evaluate(context);

  // Modify the blend params and then reevaluate.
  Scalar weights[];
  weights.resize(2);
  weights[0] = 0.65;
  weights[1] = 1.0;
  blendShapesModifier.setBlendWeights(weights);

  GeometrySet geomSet = stack.evaluate(context);

  StopFabricProfiling();

  report( GetEvalPathReport() );

  report("stack:" + stack.getDesc());

}

