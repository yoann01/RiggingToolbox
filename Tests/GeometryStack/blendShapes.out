loadTargetsFromBinCache:D:/Projects/FabricEngineInc/RiggingToolbox/Tests/GeometryStack/Resources/sphereBlendShapes.blendShapes
Importing:D:/Projects/FabricEngineInc/RiggingToolbox/Tests/GeometryStack/Resources/sphereBlendShapes.abc
BlendShapesModifier cannot evaluate because the number of weights:0 does not match the number of targets:2
numActiveShapes:0
numActiveShapes:2
function GeometrySet GeometryStack.evaluate!(EvalContext)
--function GeometryCache.update!(io GeometrySet, GeometryOperator)
--function AlembicGeometryGenerator.evaluate!(EvalContext, io GeometrySet)
--function GeometryAttributeCache.update!(io GeometrySet, GeometryOperator):["positions"]
----Update:positions
--function BlendShapesModifier.evaluate!(EvalContext, io GeometrySet)
----function blendShapesModifier_deformGeometries(Index, io GeometrySet, BlendShapesModifier_Target[][], Scalar[], Scalar, Boolean, Color[], io UInt32)
function GeometryStack.notify!(Notifier, String, String):BlendShapesModifier.changed
function GeometrySet GeometryStack.evaluate!(EvalContext)
--function GeometryAttributeCache.update!(io GeometrySet, GeometryOperator):["positions"]
----Restore:positions
--function BlendShapesModifier.evaluate!(EvalContext, io GeometrySet)
----function blendShapesModifier_deformGeometries(Index, io GeometrySet, BlendShapesModifier_Target[][], Scalar[], Scalar, Boolean, Color[], io UInt32)

stack:GeometryStack {
  geomOperators:[
    {
      type: AlembicGeometryGenerator,
      filePath: sphereBlendShapes.abc
      expandedPath: D:/Projects/FabricEngineInc/RiggingToolbox/Tests/GeometryStack/Resources/sphereBlendShapes.abc
    }
    {
      type: BlendShapesModifier,
      filePath: sphereBlendShapes.abc
      refGeoms: 1
        blendTargets: 0:2
    }
  ],
  geomSet:   {
    geometries:[
      SphereRefShape:{
      },
    ]
    attributeGenerations:{
      positions:2,
    },
  }
}
